基本設計書（初版）

yutakanakajima opened the pull request:
基本設計書を作成しました。
・検索エンドポイントは全件、個別と共通で使用する予定。
・登録エンドポイントは、DBに無い項目をパラメータで受け取る予定です。
　消費税は、今回のWebアプリでは、アプリ内で8% 固定で考えています。

yutakanakajima committed:
初版

yutakanakajima committed:
dao のみ実装

kudarar commented on invoice-project/invoice-dao/src/main/java/jp/dip/invoiceapp/app/InvoiceDao.java:
```diff
@@ -0,0 +1,223 @@
+import jp.dip.invoiceapp.app.tbl.InvoiceTblData;
+import jp.dip.invoiceapp.app.tbl.OrderTblData;
+
+public class InvoiceDao {
```
https://dev.classmethod.jp/server-side/java/use_spring-boot-jpa-jpql/
今回はカリキュラム要件として、Spring JPA を使ってDao層の実装をお願いしております。
Spring JPA を使うことでO/Rマッパーが提供されるため、
```
			while(rs.next()){
				InvoiceTblData invoice = new InvoiceTblData();
			    invoice.setInvoiceNo(rs.getInt("invoice_no"));
```
このような冗長記述が減らせるなど、コーディングがスッキリしてスピード向上と品質安定・統一を図ることが出来ます。

kudarar commented on invoice-project/invoice-dao/src/main/java/jp/dip/invoiceapp/app/MySQLConnector.java:
```diff
@@ -0,0 +1,36 @@
+import java.sql.DriverManager;
+import java.sql.SQLException;
+
+public class MySQLConnector {
```
http://javatechnology.net/spring/spring-boot-mysql-connect/
Spring Boot の機能により、DatasouceのBean生成やコネクションプールのAPIを自動作成することができます。

kudarar commented on invoice-project/invoice-dao/src/main/java/jp/dip/invoiceapp/app/tbl/ClientTblData.java:
```diff
@@ -0,0 +1,79 @@
+
+import java.util.Date;
+
+public class ClientTblData {
```
https://qiita.com/rubytomato@github/items/cb7c40cea2ff3225d5e3
JPAのEntityクラスをDBからリバースして自動生成することが出来ます。

kudarar reviewed (changes requested):
Dao層部分について現時点の内容でレビューさせていただきました。

yutakanakajima committed:
以下のレビュー指摘対応
・共通
（１）/InvoiceSearchServlet、/InvoiceEntryServlet　のパス命名に動詞が入っている。
　REST表現ですので、リソースを表現する思想となっております。
　今回は①"請求書"というリソースの対象をURIで、②操作をHTTPメソッドで表現する必要があります。
　①に関しては動詞が含まれており、RESTARTED表現ではあまり好ましくないとされております。
　（業務運用では問題は無いと思いますが、あくまで設計思想の観点です。
　したがって、 両者とも/Invoice のみとするほうが適切な設計になります。

（２）errorValue が単一想定となっている。
　特に登録のときに支障がでてくる設計になるかと思います。
　例えば複数項目のバリデーションエラーをまとめて返却できないため、
　利用者のユースケースによっては不便な仕様となります。
　一項目のString中に複数のエラーを書き込むことも好ましくないため、
　配列で複数返却できるようにして貰えればと思います。
　
（３）請求書管理番号には数値チェックを入れてもらえればと思います。

（４）プロトコルはHTTPでも大丈夫です。そのままでもよいです。
　（外部公開を想定したカリキュラム設定ではないため。

・登録エンドポイント
（１）正常時は登録後の「請求書管理番号」を返却するほうが設計として好ましいと思います。
（２）ユーザIDをリクエストパラメタに含めて貰えればと思います。
　　　登録後は登録ユーザIDと更新ユーザIDに設定されるものです。

・検索エンドポイント
　なし
